File: src/main/java/at/searles/parsing/printing/ConcreteSyntaxTree.java
package at.searles.parsing.printing;

import java.util.List;

public interface ConcreteSyntaxTree {

    ConcreteSyntaxTree EMPTY = new EmptyConcreteSyntaxTree();

    static ConcreteSyntaxTree empty() {
        return EMPTY;
    }

    static ConcreteSyntaxTree fromCharSequence(CharSequence seq) {
        return new LeafConcreteSyntaxTree(seq);
    }

    static ConcreteSyntaxTree fromList(List<ConcreteSyntaxTree> list) {
        if (list.isEmpty()) {
            return empty();
        } else if (list.size() == 1) {
            return list.get(0);
        } else if (list.size() == 2) {
            return list.get(0).consRight(list.get(1));
        } else {
            return new ListConcreteSyntaxTree(list);
        }
    }

    /**
     * Print this syntax tree using the provided printer.
     *
     * @param printer The non-null printer.
     */
    void printTo(CstPrinter printer);

    default ConcreteSyntaxTree consRight(ConcreteSyntaxTree right) {
        return new ConsConcreteSyntaxTree(this, right);
    }

    default ConcreteSyntaxTree consLeft(ConcreteSyntaxTree left) {
        return new ConsConcreteSyntaxTree(left, this);
    }

    default <C> ConcreteSyntaxTree annotate(C annotate) {
        return new AnnotatedConcreteSyntaxTree<>(annotate, this);
    }
}

File: src/main/java/at/searles/parsing/printing/ConcreteSyntaxTree.kt
package at.searles.parsing.printing

interface ConcreteSyntaxTree {
    /**
     * Print this syntax tree using the provided printer.
     *
     * @param visitor The non-null printer.
     */
    fun accept(visitor: CstVisitor)

    fun consRight(right: ConcreteSyntaxTree): ConcreteSyntaxTree {
        return ConsTree(this, right)
    }

    fun consLeft(left: ConcreteSyntaxTree): ConcreteSyntaxTree {
        return ConsTree(left, this)
    }

    companion object {
        private val EMPTY: ConcreteSyntaxTree = EmptyTree()

        fun empty(): ConcreteSyntaxTree {
            return EMPTY
        }

        fun fromCharSequence(seq: CharSequence): ConcreteSyntaxTree {
            return TokenTree(seq)
        }

        fun fromList(list: List<ConcreteSyntaxTree>): ConcreteSyntaxTree {
            return when {
                list.isEmpty() -> empty()
                list.size == 1 -> list[0]
                list.size == 2 -> list[0].consRight(list[1])
                else -> ListTree(list)
            }
        }
    }
}